production/helm/loki/templates/_helpers.tpl

```diff
@@ -245,7 +245,20 @@ s3:
     ca_file: {{ . }}
     {{- end}}
   {{- end }}
+  {{- with .backoff_config}}
```

Comment:
Can you add an if statement where to check if backoff_config is set? If it's not set/empty then this should not create anything.

Reply:
I don't think an if statement is needed here. tested this with a few different values and if the backoff_config isn't specified or is an empty object it just doesn't get added to the generated config:

1. backoff_config is set
    * values:
    ```
      storage:
        bucketNames:
          chunks: loki-chunks
          ruler: loki-ruler
          admin: loki-admin
        type: s3
        s3:
          endpoint: https://some.hostname.com:443
          accessKeyId: ACCESS_KEY
          secretAccessKey: SECRET_KEY
          s3ForcePathStyle: false
          insecure: false
          backoff_config:
            min_period: "2s"
            max_period: "5s"
            max_retries: 5
    ```

    * resulting s3 configuration:
    ```
      storage:
        s3:
          access_key_id: ACCESS_KEY
          backoff_config:
            max_period: 5s
            max_retries: 5
            min_period: 2s
          bucketnames: loki-chunks
          endpoint: https://some.hostname.com:443
          insecure: false
          s3forcepathstyle: false
          secret_access_key: SECRET_KEY
    ```

2. backoff_config with empty object:
    * values:
    ```
      storage:
        bucketNames:
          chunks: loki-chunks
          ruler: loki-ruler
          admin: loki-admin
        type: s3
        s3:
          endpoint: https://some.hostname.com:443
          accessKeyId: ACCESS_KEY
          secretAccessKey: SECRET_KEY
          s3ForcePathStyle: false
          insecure: false
          backoff_config: {}
    ```

    * resulting s3 configuration:
    ```
      storage:
        s3:
          access_key_id: ACCESS_KEY
          bucketnames: loki-chunks
          endpoint: https://some.hostname.com:443
          insecure: false
          s3forcepathstyle: false
          secret_access_key: SECRET_KEY
    ```
2. backoff_config not specified:
    * values:
    ```
      storage:
        bucketNames:
          chunks: loki-chunks
          ruler: loki-ruler
          admin: loki-admin
        type: s3
        s3:
          endpoint: https://some.hostname.com:443
          accessKeyId: ACCESS_KEY
          secretAccessKey: SECRET_KEY
          s3ForcePathStyle: false
          insecure: false
    ```

    * resulting s3 configuration:
    ```
      storage:
        s3:
          access_key_id: ACCESS_KEY
          bucketnames: loki-chunks
          endpoint: https://some.hostname.com:443
          insecure: false
          s3forcepathstyle: false
          secret_access_key: SECRET_KEY
    ```

Reply:
You are right. I should have read the docs for the with statement:
```
{{with pipeline}} T1 {{end}}
	If the value of the pipeline is empty, no output is generated;
	otherwise, dot is set to the value of the pipeline and T1 is
	executed.
```